crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -27,29 +27,115 @@ use {
         error::Math,
         trade::{ClearingPrices, Fee, Fulfillment},
     },
-    crate::domain::{
-        competition::{
-            order,
-            order::{FeePolicy, Side},
-            solution::error::Trade,
-            PriceLimits,
+    crate::{
+        domain::{
+            competition::{
+                order::{self, FeePolicy, Side, TargetAmount},
+                solution::{
+                    error::{self, Trade as TradeError},
+                    trade::{Jit, Trade},
+                },
+                PriceLimits,
+            },
+            eth::{self, Asset},
         },
-        eth::{self},
+        util::conv::u256::U256Ext,
     },
     bigdecimal::Zero,
 };
 
-impl Fulfillment {
+impl Trade {
```

Comment:
My idea was to use the full `Trade` to calculate the protocol fees, independently of if it is a `Jit` or `Fulfillment` order, this was it should be very straightforward to support more fee policies for either type of order.